### ISSUE TITLE
Can't rename column after read_csv when files have different column names

### ISSUE THREAD

**Issue by Puumanamana** [dask/dask#10198]
**Describe the issue**: When reading csv files with different column names `dask.dataframe.DataFrame.read_csv()`, renaming a column (shared by all files) before computing throws an error. This error doesn't occur if I don't do `.rename()` 

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe as dd

data = dd.read_csv("file*.csv")
data.compute() # works fine
data.rename(columns=dict(c1="c_new")).compute() # error
```

```bash
$ cat file1.csv
c1,c2
1,1

$ cat file2.csv
c1,c3
1,1
```
Error:
```
ValueError: The columns in the computed data do not match the columns in the provided metadata
  Extra:   ['c3']
  Missing: ['c2']
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2023.3.0
- Python version: 3.11.0
- Operating System: Debian GNU/Linux 10
- Install method (conda, pip, source): conda (mamba)

**Comment by rjzamora** [comment]
Thanks for raising @Puumanamana 

The `csv` reader doesn't really handle the case that there is an inconsistent schema across different files in the dataset. The problem is not really related to `rename`, but to the fact that any default `map_partitions` call will raise an error if the the columns produced by a task are different from the columns in the global `_meta` attribute (and in your case the `_meta` attribute will only have the columns found in the first file).

I'm not sure if there is a simple way to deal with this limitation besides allowing the user to pass in a custom `meta` argument.

If you are okay with having a 1:1 mapping between files and partitions, `from_map` provide a work-around:

```python
import dask.dataframe as dd
import pandas as pd

# Define the schema you want to enforce
meta = pd.DataFrame({k: pd.Series([], dtype="Int64") for k in ["c1", "c2", "c3"]})

def read_csv_file(path):
    # Read in a file and enforce the global schema
    return pd.concat([meta, pd.read_csv(path)])

data = dd.from_map(read_csv_file, ["file0.csv", "file1.csv"], meta=meta)
data.rename(columns=dict(c1="c_new")).compute()
```
```
   c_new    c2    c3
0      1     1  <NA>
0      1  <NA>     1
```

**Comment by Puumanamana** [comment]
I see, that makes sense. For my use case, I'm just doing some preprocessing to make all dataframes uniform (with the same schema) before combining them. 
I think your solution would work for me. Would there be any performance difference between that and dd.read_csv? Also, maybe another alternative would be to start from a bag, and then cast to a dataframe?

**Comment by rjzamora** [comment]
>For my use case, I'm just doing some preprocessing to make all dataframes uniform (with the same schema) before combining them.

Right. This is probably the ideal approach. I suppose you can also do this "pre-processing" after the `read_csv` call using `map_partitions`:

```python
meta = pd.DataFrame({k: pd.Series([], dtype="Int64") for k in ["c1", "c2", "c3"]})
data = dd.read_csv("file*csv").map_partitions(lambda x: pd.concat([meta, x]))
```
(or replace `pd.concat` with some other logic to enforce the schema you want)

> I think your solution would work for me. Would there be any performance difference between that and dd.read_csv?

`from_map` is quite performant, but using `map_partitions` after `dd.read_csv` is probably better (sorry for missing that earlier).

>Also, maybe another alternative would be to start from a bag, and then cast to a dataframe?

I'd probably avoid using bag here if it isn't necessary.
